File: docx_parser_converter/docx_to_html/converters/table_converter.py
from typing import List, Optional, Sequence, cast
from lxml import etree, html  # type: ignore
from docx_parser_converter.docx_parsers.models.table_models import (
    MarginProperties,
    Table,
    TableCell,
    TableCellProperties,
    TableProperties,
    TableRow,
    TableRowProperties,
)
from docx_parser_converter.docx_to_html.converters.paragraph_converter import (
    ParagraphConverter,
)
from docx_parser_converter.docx_to_html.converters.style_converter import (
    StyleConverter,
)

class TableConverter:
    """
    A converter class for converting DOCX tables to HTML.
    """

    @staticmethod
    def convert_table(table: Table) -> str:
        """
        Converts a table to its HTML representation.

        Args:
            table (Table): The table to convert.

        Returns:
            str: The HTML representation of the table.

        Example:
            Given a table with properties, the output HTML string might look like:

            .. code-block:: html

                <table style="border-collapse: collapse; width:500pt;">
                    <colgroup>
                        <col style="width:250pt;">
                        <col style="width:250pt;">
                    </colgroup>
                    <tbody>
                        <tr style="height:20pt;">
                            <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 1</td>
                            <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 2</td>
                        </tr>
                    </tbody>
                </table>
        """
        table_html = etree.Element("table")
        properties = table.properties or TableProperties.model_validate({})
        table_properties_html = TableConverter.convert_table_properties(properties)
        table_html.set("style", table_properties_html)

        if table.grid and table.grid.columns:
            table_grid_html = TableConverter.convert_grid(table.grid.columns)
            table_html.append(html.fragment_fromstring(table_grid_html))

        rows_html = TableConverter.convert_rows(table.rows, properties.tblCellMar)
        table_html.append(rows_html)

        return cast(
            str, html.tostring(table_html, pretty_print=True, encoding="unicode")
        )

    @staticmethod
    def convert_table_properties(properties: Optional[TableProperties]) -> str:
        """
        Converts table properties to an HTML style attribute.

        Args:
            properties: The table properties to convert.

        Returns:
            str: The HTML style attribute representing the table properties.

        Example:
            The output style attribute might look like:

            .. code-block:: css

                'border-collapse: collapse; width:500pt; text-align:center; margin-left:20pt; padding: 5pt 5pt 5pt 5pt;'
        """
        props = properties or TableProperties.model_validate({})
        styles = ["border-collapse: collapse;"]  # Ensure borders collapse
        if props.tblW and props.tblW.width is not None:
            styles.append(f"width:{props.tblW.width}pt;")
        if props.justification:
            justification_style = StyleConverter.convert_justification(
                props.justification
            )
            if justification_style:
                styles.append(justification_style)
        if props.tblInd and props.tblInd.width is not None:
            styles.append(f"margin-left:{props.tblInd.width}pt;")
        if props.tblCellMar:
            styles.append(
                f"padding: {props.tblCellMar.top}pt {props.tblCellMar.right}pt "
                f"{props.tblCellMar.bottom}pt {props.tblCellMar.left}pt;"
            )
        if props.tblLayout:
            styles.append(f"table-layout:{props.tblLayout};")
        return " ".join(styles)

    @staticmethod
    def convert_grid(columns: Sequence[float]) -> str:
        """
        Converts table grid columns to HTML.

        Args:
            columns: The grid columns widths.

        Returns:
            str: The HTML colgroup element representing the grid columns.

        Example:
            The output HTML colgroup might look like:

            .. code-block:: html

                <colgroup>
                    <col style="width:250pt;">
                    <col style="width:250pt;">
                </colgroup>
        """
        colgroup = etree.Element("colgroup")
        for width in columns:
            etree.SubElement(colgroup, "col", style=f"width:{width}pt;")
        return cast(
            str, html.tostring(colgroup, pretty_print=True, encoding="unicode")
        )

    @staticmethod
    def convert_rows(
        rows: Sequence[TableRow], tblCellMar: Optional[MarginProperties]
    ) -> etree.Element:
        """
        Converts table rows to HTML.

        Args:
            rows: The list of table rows to convert.
            tblCellMar: The table cell margins to apply.

        Returns:
            etree.Element: The HTML tbody element representing the rows.

        Example:
            The output HTML tbody might look like:

            .. code-block:: html

                <tbody>
                    <tr style="height:20pt;">
                        <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 1</td>
                        <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 2</td>
                    </tr>
                </tbody>
        """
        tbody = etree.Element("tbody")
        for row in rows:
            row_html = TableConverter.convert_row(row, tblCellMar)
            tbody.append(row_html)
        return tbody

    @staticmethod
    def convert_row(row: TableRow, tblCellMar: Optional[MarginProperties]) -> etree.Element:
        """
        Converts a table row to HTML.

        Args:
            row: The row to convert.
            tblCellMar: The table cell margins to apply.

        Returns:
            etree.Element: The HTML tr element representing the row.

        Example:
            The output HTML tr might look like:

            .. code-block:: html

                <tr style="height:20pt;">
                    <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 1</td>
                    <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 2</td>
                </tr>
        """
        tr = etree.Element("tr")
        row_properties_html = TableConverter.convert_row_properties(row.properties)
        if row_properties_html:
            tr.set("style", row_properties_html)

        cells_html = TableConverter.convert_cells(row.cells, tblCellMar)
        for cell_html in cells_html:
            tr.append(cell_html)
        
        return tr

    @staticmethod
    def convert_row_properties(
        properties: Optional[TableRowProperties],
    ) -> str:
        """
        Converts row properties to an HTML style attribute.

        Args:
            properties: The row properties to convert.

        Returns:
            str: The HTML style attribute representing the row properties.

        Example:
            The output style attribute might look like:

            .. code-block:: css

                'height:20pt; font-weight:bold;'
        """
        row_props = properties or TableRowProperties.model_validate({})
        styles: List[str] = []
        if row_props.trHeight:
            tr_height = (
                float(row_props.trHeight)
                if isinstance(row_props.trHeight, str)
                else row_props.trHeight
            )
            if tr_height is not None:
                styles.append(f"height:{tr_height}pt;")
        if row_props.tblHeader:
            styles.append("font-weight:bold;")
        return " ".join(styles)

    @staticmethod
    def convert_cells(
        cells: Sequence[TableCell], tblCellMar: Optional[MarginProperties]
    ) -> list:
        """
        Converts table cells to HTML.

        Args:
            cells: The list of cells to convert.
            tblCellMar: The table cell margins to apply.

        Returns:
            list: The list of HTML td elements representing the cells.

        Example:
            The output HTML td elements might look like:

            .. code-block:: html

                <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 1</td>
                <td style="width:250pt; padding:5pt 5pt 5pt 5pt;">Cell 2</td>
        """
        cell_elements = []
        for cell in cells:
            td = etree.Element("td")
            cell_properties_html = TableConverter.convert_cell_properties(cell.properties, tblCellMar)
            td.set("style", cell_properties_html)
            
            if TableConverter.is_cell_empty(cell):  # Check if the cell is empty
                td.text = "\u00A0"
            else:
                for paragraph in cell.paragraphs:
                    paragraph_html = ParagraphConverter.convert_paragraph(paragraph, None)
                    td.append(html.fragment_fromstring(paragraph_html))
            
            cell_elements.append(td)
        return cell_elements

    @staticmethod
    def is_cell_empty(cell) -> bool:
        """
        Check if a cell is empty by verifying if all runs in all paragraphs have no contents.

        Args:
            cell: The cell to check.

        Returns:
            bool: True if the cell is empty, False otherwise.

        Example:
            .. code-block:: python

                if TableConverter.is_cell_empty(cell):
                    print("Cell is empty")
                else:
                    print("Cell is not empty")
        """
        for paragraph in cell.paragraphs:
            for run in paragraph.runs:
                if run.contents:
                    return False
        return True

    @staticmethod
    def convert_cell_properties(properties, tblCellMar) -> str:
        """
        Converts cell properties to an HTML style attribute.

        Args:
            properties: The cell properties to convert.
            tblCellMar: The table cell margins to apply.

        Returns:
            str: The HTML style attribute representing the cell properties.

        Defaults:
            - Word wrapping is enabled with "word-wrap: break-word;" and "word-break: break-all;".
            - Overflow handling is enabled with "overflow-wrap: break-word;" and "overflow: hidden;".
            - Vertical alignment defaults to "top".

        Example:
            The output style attribute might look like:

            .. code-block:: css

                'width:250pt; border-top:1pt solid #000000; border-left:1pt solid #000000; border-bottom:1pt solid #000000; border-right:1pt solid #000000; background-color:#FFFFFF; padding:5pt 5pt 5pt 5pt; vertical-align:top;'
        """
        cell_properties = properties or TableCellProperties.model_validate({})
        styles = [
            "word-wrap: break-word;",       # Allow words to be broken at arbitrary points
            "word-break: break-all;",       # Ensure long words break and wrap into next line
            "overflow-wrap: break-word;",   # Handle long words in tables
            "overflow: hidden;"             # Hide overflow content
        ]
        if cell_properties.tcW and cell_properties.tcW.width is not None:
            styles.append(f"width:{cell_properties.tcW.width}pt;")
        if cell_properties.tcBorders:
            borders = cell_properties.tcBorders
            if borders.top and borders.top.size is not None:
                styles.append(
                    f"border-top:{borders.top.size / 8}pt "
                    f"{TableConverter.map_border_style(borders.top.val)} #{borders.top.color};"
                )
            if borders.left and borders.left.size is not None:
                styles.append(
                    f"border-left:{borders.left.size / 8}pt "
                    f"{TableConverter.map_border_style(borders.left.val)} #{borders.left.color};"
                )
            if borders.bottom and borders.bottom.size is not None:
                styles.append(
                    f"border-bottom:{borders.bottom.size / 8}pt "
                    f"{TableConverter.map_border_style(borders.bottom.val)} #{borders.bottom.color};"
                )
            if borders.right and borders.right.size is not None:
                styles.append(
                    f"border-right:{borders.right.size / 8}pt "
                    f"{TableConverter.map_border_style(borders.right.val)} #{borders.right.color};"
                )
        if cell_properties.shd and cell_properties.shd.fill:
            styles.append(f"background-color:#{cell_properties.shd.fill};")
        if tblCellMar:
            styles.append(
                f"padding: {tblCellMar.top}pt {tblCellMar.right}pt "
                f"{tblCellMar.bottom}pt {tblCellMar.left}pt;"
            )
        alignment_value = getattr(cell_properties, "verticalAlignment", None)
        if not alignment_value and cell_properties.vAlign:
            alignment_value = cell_properties.vAlign
        if alignment_value:
            styles.append(
                f"vertical-align:{TableConverter.map_vertical_alignment(alignment_value)};"
            )
        else:
            styles.append("vertical-align:top;")

        text_alignment = getattr(cell_properties, "textAlignment", None)
        if text_alignment:
            styles.append(f"text-align:{text_alignment};")
        return " ".join(styles)

    @staticmethod
    def map_border_style(val) -> str:
        """
        Maps DOCX border style to CSS border style.

        Args:
            val: The DOCX border style.

        Returns:
            str: The CSS border style.

        Example:
            The output CSS border style might look like:

            .. code-block:: css

                'solid'
        """
        mapping = {
            "single": "solid",
            "double": "double",
            "dashed": "dashed",
            # Add more mappings as needed
        }
        return mapping.get(val, "solid")  # Default to solid if style is unknown

    @staticmethod
    def map_vertical_alignment(val) -> str:
        """
        Maps DOCX vertical alignment to CSS vertical alignment.

        Args:
            val: The DOCX vertical alignment.

        Returns:
            str: The CSS vertical alignment.

        Example:
            The output CSS vertical alignment might look like:

            .. code-block:: css

                'top'
        """
        mapping = {
            "top": "top",
            "center": "middle",
            "bottom": "bottom",
            # Add more mappings as needed
        }
        return mapping.get(val, "top")  # Default to top if alignment is unknown

File: docx_parser_converter_ts/src/docx_to_html/converters/tableConverter.ts
import type { 
    Table, TableProperties, TableGrid, TableRow, TableRowProperties, 
    TableCell, TableCellProperties, MarginProperties
} from '../../docx_parsers/models/tableModels';
import type { NumberingSchema } from '../../docx_parsers/models/numberingModels';
import { ParagraphConverter } from './paragraphConverter';

export class TableConverter {

    private static readonly EMPTY_NUMBERING_SCHEMA: NumberingSchema = { instances: [] };

    public static convertTable(table: Table): string {
        const tablePropertiesStyle = TableConverter.convertTableProperties(table.properties);
        let tableHtml = `<table style="${tablePropertiesStyle}">`;

        if (table.grid) {
            tableHtml += TableConverter.convertGrid(table.grid);
        }
        
        tableHtml += TableConverter.convertRows(table.rows, table.properties?.tblCellMar, TableConverter.EMPTY_NUMBERING_SCHEMA);
        
        tableHtml += "</table>";
        return tableHtml;
    }

    public static convertTableProperties(properties?: TableProperties): string {
        const styles: string[] = ["border-collapse: collapse;"];
        if (!properties) return styles.join(" ");

        if (properties.tblW?.width !== undefined) {
            styles.push(`width:${properties.tblW.width}pt;`);
        }
        if (properties.justification) {
            styles.push(`text-align:${properties.justification};`);
        }
        if (properties.tblInd?.width !== undefined) {
            styles.push(`margin-left:${properties.tblInd.width}pt;`);
        }
        // Note: Python version applies tblCellMar padding to the table style itself.
        // This is unusual; cell padding is typically on <td> or <th>.
        // Replicating Python logic:
        if (properties.tblCellMar) {
            const mar = properties.tblCellMar;
            styles.push(`padding: ${mar.top ?? 0}pt ${mar.right ?? 0}pt ${mar.bottom ?? 0}pt ${mar.left ?? 0}pt;`);
        }
        if (properties.tblLayout) {
            styles.push(`table-layout:${properties.tblLayout};`);
        }
        return styles.join(" ");
    }

    public static convertGrid(grid: TableGrid): string {
        let colgroupHtml = "<colgroup>";
        for (const width of grid.columns) {
            colgroupHtml += `<col style="width:${width}pt;">`;
        }
        colgroupHtml += "</colgroup>";
        return colgroupHtml;
    }

    public static convertRows(rows: TableRow[], tblCellMar?: MarginProperties, numberingSchema?: NumberingSchema): string {
        let tbodyHtml = "<tbody>";
        for (const row of rows) {
            tbodyHtml += TableConverter.convertRow(row, tblCellMar, numberingSchema ?? TableConverter.EMPTY_NUMBERING_SCHEMA);
        }
        tbodyHtml += "</tbody>";
        return tbodyHtml;
    }

    public static convertRow(row: TableRow, tblCellMar?: MarginProperties, numberingSchema?: NumberingSchema): string {
        const rowPropertiesStyle = TableConverter.convertRowProperties(row.properties);
        let trHtml = `<tr style="${rowPropertiesStyle}">`;
        trHtml += TableConverter.convertCells(row.cells, tblCellMar, numberingSchema ?? TableConverter.EMPTY_NUMBERING_SCHEMA);
        trHtml += "</tr>";
        return trHtml;
    }

    public static convertRowProperties(properties?: TableRowProperties): string {
        const styles: string[] = [];
        if (!properties) return "";

        if (properties.trHeight) {
            const height = parseFloat(properties.trHeight); // In points
            if (!isNaN(height)) {
                styles.push(`height:${height}pt;`);
            }
        }
        if (properties.tblHeader) {
            styles.push("font-weight:bold;");
        }
        return styles.join(" ");
    }

    public static convertCells(cells: TableCell[], tblCellMar?: MarginProperties, numberingSchema?: NumberingSchema): string {
        let cellsHtml = "";
        for (const cell of cells) {
            const cellPropertiesStyle = TableConverter.convertCellProperties(cell.properties, tblCellMar);
            cellsHtml += `<td style="${cellPropertiesStyle}">`;
            
            if (TableConverter.isCellEmpty(cell)) {
                cellsHtml += "&nbsp;"; // Non-breaking space for empty cells
            } else {
                for (const paragraph of cell.paragraphs) {
                    cellsHtml += ParagraphConverter.convertParagraph(paragraph, numberingSchema ?? TableConverter.EMPTY_NUMBERING_SCHEMA);
                }
            }
            cellsHtml += "</td>";
        }
        return cellsHtml;
    }

    public static isCellEmpty(cell: TableCell): boolean {
        if (!cell.paragraphs || cell.paragraphs.length === 0) return true;
        for (const paragraph of cell.paragraphs) {
            if (paragraph.runs && paragraph.runs.length > 0) {
                for (const run of paragraph.runs) {
                    if (run.contents && run.contents.length > 0) {
                        // Check if any content is actual text and not just empty objects or tabs
                        const hasText = run.contents.some(content => content.type === 'text' && content.text.trim() !== '');
                        if(hasText) return false;
                        // If it contains only tabs or empty text, it might still be considered empty for rendering purposes
                        // However, original python checks for any content. So if contents array is not empty, it is not empty.
                        // Replicating python: if run.contents: return False
                        if(run.contents.length > 0) return false;
                    }
                }
            }
        }
        return true;
    }

    public static convertCellProperties(properties?: TableCellProperties, tblCellMar?: MarginProperties): string {
        const styles: string[] = [
            "word-wrap: break-word;",
            "word-break: break-all;", 
            "overflow-wrap: break-word;",
            "overflow: hidden;" 
        ];
        
        if (!properties && !tblCellMar) return styles.join(" "); // Minimal default styles

        if (properties?.tcW?.width !== undefined) {
            styles.push(`width:${properties.tcW.width}pt;`);
        }
        if (properties?.tcBorders) {
            const { top, left, bottom, right } = properties.tcBorders;
            if (top?.val) {
                styles.push(`border-top:${(top.size ?? 0) / 8}pt ${TableConverter.mapBorderStyle(top.val)} #${top.color || '000000'};`);
            }
            if (left?.val) {
                styles.push(`border-left:${(left.size ?? 0) / 8}pt ${TableConverter.mapBorderStyle(left.val)} #${left.color || '000000'};`);
            }
            if (bottom?.val) {
                styles.push(`border-bottom:${(bottom.size ?? 0) / 8}pt ${TableConverter.mapBorderStyle(bottom.val)} #${bottom.color || '000000'};`);
            }
            if (right?.val) {
                styles.push(`border-right:${(right.size ?? 0) / 8}pt ${TableConverter.mapBorderStyle(right.val)} #${right.color || '000000'};`);
            }
        }
        if (properties?.shd?.fill) {
            styles.push(`background-color:#${properties.shd.fill};`);
        }
        
        // Cell-specific margins (tcMar) take precedence over table-wide cell margins (tblCellMar)
        const cellMarginToApply = properties?.tcMar || tblCellMar;
        if (cellMarginToApply) {
            const mar = cellMarginToApply;
            styles.push(`padding: ${mar.top ?? 0}pt ${mar.right ?? 0}pt ${mar.bottom ?? 0}pt ${mar.left ?? 0}pt;`);
        }

        if (properties?.vAlign) {
            styles.push(`vertical-align:${TableConverter.mapVerticalAlignment(properties.vAlign)};`);
        } else {
            styles.push("vertical-align: top;"); // Default
        }
        
        // textAlignment on cell properties is not in the TS model TableCellProperties.
        // If needed, it should be handled by paragraph properties inside the cell.
        // Python script: if hasattr(properties, 'textAlignment'): styles.append(f"text-align:{properties.textAlignment};")
        // This is omitted as it's not in the model.

        return styles.join(" ");
    }

    public static mapBorderStyle(val?: string): string {
        if (!val) return "solid"; // Default
        const mapping: Record<string, string> = {
            "single": "solid",
            "double": "double",
            "dashed": "dashed",
            // TODO: Add more DOCX border styles to CSS mappings if needed
            // e.g., dotted, dashDot, dashDotDot, etc.
        };
        return mapping[val.toLowerCase()] || "solid";
    }

    public static mapVerticalAlignment(val?: string): string {
        if (!val) return "top"; // Default
        const mapping: Record<string, string> = {
            "top": "top",
            "center": "middle",
            "bottom": "bottom",
            // "both" (justify) is not a standard CSS vertical-align value for table cells directly.
        };
        return mapping[val.toLowerCase()] || "top";
    }
}

// Example Usage (equivalent to Python's if __name__ == "__main__")
/*
import { RunConverter } from './runConverter'; // For NumberingConverter.resetCounters if used in test setup for paragraphs
import { NumberingConverter } from './numberingConverter'; // For NumberingConverter.resetCounters

function testTableConverter() {
    console.log("--- Testing TableConverter ---");
    NumberingConverter.resetCounters(); // Reset for any paragraph numbering within cells

    // --- Mock Data Setup ---
    const sampleTable: Table = {
        properties: {
            tblW: { width: 500, type: "pt" },
            justification: "center",
            tblCellMar: { top: 5, right: 5, bottom: 5, left: 5 }, // Default cell margins for the table
            tblLayout: "fixed",
            tblBorders: { // Table-level borders (usually for outer borders)
                top: { val: "single", size: 8, color: "0000FF" }, // Blue top border for table
                bottom: { val: "single", size: 8, color: "0000FF" },
            }
        },
        grid: {
            columns: [150, 200, 150] // Three columns
        },
        rows: [
            {
                properties: { trHeight: "20", tblHeader: true }, // Row height in points, is a header row
                cells: [
                    {
                        properties: { tcW: { width: 150 }, vAlign: "center" }, 
                        paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: "Header 1" }] }] }]
                    },
                    {
                        properties: { tcW: { width: 200 }, tcBorders: { bottom: { val: "dashed", size: 16, color: "FF0000" } } }, 
                        paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: "Header 2 (Red Dashed Bottom)" }] }] }]
                    },
                    {
                        properties: { tcW: { width: 150 }, shd: { fill: "DDDDDD" } }, // Shaded cell
                        paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: "Header 3 (Shaded)" }] }] }]
                    }
                ]
            },
            {
                cells: [
                    {
                        // Cell with specific margin overriding table default
                        properties: { tcMar: { top: 10, right: 10, bottom: 10, left: 10 } }, 
                        paragraphs: [
                            { properties: { justification: "right" }, runs: [{ contents: [{ type: 'text', text: "Row 2, Cell 1 (Right Aligned Para)" }] }] },
                            { properties: {}, runs: [{ contents: [{ type: 'text', text: "Second para in cell"}]}]}
                        ]
                    },
                    {
                        paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: "Row 2, Cell 2" }] }] }]
                    },
                    { 
                        // Empty cell test
                        paragraphs: [{ properties: {}, runs: [{ contents: []}]}]
                    }
                ]
            }
        ]
    };

    // --- Test Execution ---
    console.log("\n--- Full Table HTML ---");
    const htmlOutput = TableConverter.convertTable(sampleTable);
    console.log(htmlOutput);
    // Manual inspection of the complex HTML output is needed here.
    // Key things to check based on mock:
    // - Table style: border-collapse, width:500pt, text-align:center, padding for tblCellMar, table-layout:fixed
    // - Colgroup: <col style="width:150pt;"><col style="width:200pt;"><col style="width:150pt;">
    // - Row 1: style="height:20pt; font-weight:bold;"
    //   - Cell 1.1: style includes width:150pt, default padding from tblCellMar, vertical-align:middle.
    //   - Cell 1.2: style includes width:200pt, border-bottom:2pt dashed #FF0000.
    //   - Cell 1.3: style includes width:150pt, background-color:#DDDDDD.
    // - Row 2:
    //   - Cell 2.1: style includes padding:10pt 10pt 10pt 10pt. Paragraph inside should be <p style="text-align:right;">...
    //   - Cell 2.3: should contain &nbsp;

    console.log("\n--- Testing isCellEmpty ---");
    const emptyCell: TableCell = { paragraphs: [{ properties: {}, runs: [{ contents: [] }] }] };
    console.log(`isCellEmpty for empty cell: ${TableConverter.isCellEmpty(emptyCell)} (Expected: true)`);
    const nonEmptyCell: TableCell = { paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: "Hi" }] }] }] };
    console.log(`isCellEmpty for non-empty cell: ${TableConverter.isCellEmpty(nonEmptyCell)} (Expected: false)`);
    const cellWithEmptyText: TableCell = { paragraphs: [{ properties: {}, runs: [{ contents: [{ type: 'text', text: " " }] }] }] };
    // Current isCellEmpty logic (replicating python `if run.contents: return False`) will consider this non-empty if text node exists, even if whitespace.
    // If stricter definition of empty (e.g. no visible characters) is needed, isCellEmpty logic would need adjustment.
    console.log(`isCellEmpty for cell with whitespace text: ${TableConverter.isCellEmpty(cellWithEmptyText)} (Expected based on strict content check: false)`);

    console.log("\n--- Testing Specific Property Converters ---");
    const tblProps: TableProperties = { tblW: { width: 300 }, justification: "left" };
    console.log(`convertTableProperties: "${TableConverter.convertTableProperties(tblProps)}"`);

    const rowProps: TableRowProperties = { trHeight: "50.5", tblHeader: true };
    console.log(`convertRowProperties: "${TableConverter.convertRowProperties(rowProps)}"`);

    const cellProps: TableCellProperties = {
        tcW: { width: 100 },
        tcBorders: { top: { val: "double", size: 4, color: "ABCDEF" } },
        shd: { fill: "EEEEEE" },
        vAlign: "bottom"
    };
    console.log(`convertCellProperties: "${TableConverter.convertCellProperties(cellProps, { top:3,left:3,bottom:3,right:3 })}"`);
}

// To run the test:
// 1. Ensure ts-node is installed.
// 2. Ensure other converter files (ParagraphConverter.ts) are available.
// 3. All model types must be correctly defined and imported.
// 4. Uncomment the line below and run (e.g., `ts-node path/to/tableConverter.ts`)
// testTableConverter();
*/ 